cmd/xl-zones.go

```diff
@@ -0,0 +1,1085 @@
+/*
+ * MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"context"
+	"io"
+	"net/http"
+	"sort"
+	"strings"
+
+	xhttp "github.com/minio/minio/cmd/http"
+	"github.com/minio/minio/pkg/lifecycle"
+	"github.com/minio/minio/pkg/madmin"
+	"github.com/minio/minio/pkg/policy"
+	"github.com/minio/minio/pkg/sync/errgroup"
+)
+
+type xlZones struct {
+	zones []*xlSets
+}
+
+func (z *xlZones) SingleZone() bool {
+	return len(z.zones) == 1
+}
+
+// Initialize new zone of erasure codes.
+func newXLZones(endpointsList EndpointLists, formats []*formatXLV3) (ObjectLayer, error) {
+	z := &xlZones{}
+	for i, ep := range endpointsList {
+		sets, err := newXLSets(ep.Endpoints, formats[i], ep.SetCount, ep.DrivesPerSet)
+		if err != nil {
+			return nil, err
+		}
+		z.zones = append(z.zones, sets)
+	}
+	return z, nil
+}
+
+// AvailableZoneSlice attaches the methods of Uint64erface to []uint64, sorting in increasing order.
+type AvailableZoneSlice []struct {
+	Index     int
+	Available uint64
+}
+
+func (p AvailableZoneSlice) Len() int           { return len(p) }
+func (p AvailableZoneSlice) Less(i, j int) bool { return p[i].Available < p[j].Available }
+func (p AvailableZoneSlice) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+
+func (z *xlZones) getLeastUsageZoneSlice(ctx context.Context) AvailableZoneSlice {
+	var availableLayers AvailableZoneSlice
+	for i, objLayer := range z.zones {
+		info := objLayer.StorageInfo(ctx)
+		var available uint64
+		for _, davailable := range info.Available {
+			available += davailable
+		}
+		availableLayers = append(availableLayers, struct {
+			Index     int
+			Available uint64
+		}{
+			Index:     i,
+			Available: available,
+		})
+	}
+	sort.Sort(availableLayers)
+	return availableLayers
+}
+
+func (z *xlZones) getLeastUsageZone(ctx context.Context) int {
```

Comment:
In terms of overall design, I don't like that the upload target seems to be 100% deterministic, meaning that everything would be directed to some servers.

How about making the upload depend on the ratio of available space:

```
func (a AvailableZoneSlice) TotalAvailable() uint64 {
	total := uint64(0)
	for _, z := range a {
		total += z.Available
	}
	return total
}

func (z *xlZones) getUploadZoneIdx(ctx context.Context) int {
	zones := z.getLeastUsageZoneSlice(ctx)
	total := zones.TotalAvailable()
	if total == 0 {
		// Houston, we have a problem, maybe panic??
		return zones[0].Index
	}
	// choose when we reach this many
	choose := rand.Uint64() % total
	atTotal := uint64(0)
	for _, zone := range zones {
		atTotal += zone.Available
		if atTotal <= choose {
			return zone.Index
		}
	}
	// Should not happen, but print values just in case.
	panic(fmt.Errorf("reached end of zones (total: %v, atTotal: %v, choose: %v)", total, atTotal, choose))
}
```
This will make it more likely to chose a zone the more space is available. Also technically the slice doesn't need to be sorted.

Reply:
This is indeed a good idea @klauspost